orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/DefaultVariableAssignmentTransform.java

```diff
@@ -59,6 +60,36 @@ public void visitPipelineTemplate(PipelineTemplate pipelineTemplate) {
       throw new IllegalTemplateConfigurationException("Missing variable values for: " + StringUtils.join(missingVariables, ", "));
     }
 
-    // TODO rz - validate variable values match the defined variable type
+    // collect variables where value type doesn't match the required type
+    List<String> wrongTypeErrorMessages = pipelineTemplateVariables.stream()
+      .filter(templateVar -> {
+        if (missingVariables.contains(templateVar.getName())) {
+          return false; // filter missing values from this list
+        }
+
+        String expectedType = templateVar.getType();
+        if (expectedType.equals("object")) {
+          return false; // not invalid, all classes are objects
+        }
+
+        String actualType = configVars.get(templateVar.getName()).getClass().getSimpleName();
+
+        if (expectedType.equals("int") && actualType.equals("Integer")){
+          return false;
+        } else if (expectedType.equals("list") && actualType.equals("ArrayList")) {
+          return false;
+        } else if (expectedType.equals(actualType.toLowerCase())) {
+          return false;
+        } else if (expectedType.equals(actualType)){
+          return false; // types match
+        }
+        return true; //types don't match, add to list
+      })
+       .map(var -> var.getName() + " (expected type '" + var.getType() + "' found type '" + configVars.get(var.getName()).getClass().getSimpleName() + "')")
```

Comment:
We've already got the actualType, we can use that here rather than getting again.

Reply:
Can we access variables from filter in a map stage? They're local variables.